osu.Game.Rulesets.Mania/UI/ManiaPlayfield.cs

```diff
@@ -1,153 +1,95 @@
-                                    Direction = FillDirection.Horizontal,
-                                    Padding = new MarginPadding { Left = 1, Right = 1 },
-                                    Spacing = new Vector2(1, 0)
-                                },
-                            }
-                        },
-                        new Container
-                        {
-                            Name = "Barlines mask",
-                            Anchor = Anchor.TopCentre,
-                            Origin = Anchor.TopCentre,
-                            RelativeSizeAxes = Axes.Y,
-                            Width = 1366, // Bar lines should only be masked on the vertical axis
-                            BypassAutoSizeAxes = Axes.Both,
-                            Masking = true,
-                            Child = content = new Container
-                            {
-                                Name = "Bar lines",
-                                Anchor = Anchor.TopCentre,
-                                Origin = Anchor.TopCentre,
-                                RelativeSizeAxes = Axes.Y,
-                                Padding = new MarginPadding { Top = HIT_TARGET_POSITION }
-                            }
-                        },
-                        judgements = new Container<DrawableManiaJudgement>
-                        {
-                            Anchor = Anchor.TopCentre,
-                            Origin = Anchor.Centre,
-                            AutoSizeAxes = Axes.Both,
-                            Y = HIT_TARGET_POSITION + 150,
-                            BypassAutoSizeAxes = Axes.Both
-                        },
-                        topLevelContainer = new Container { RelativeSizeAxes = Axes.Both }
-                    }
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Spacing = new Vector2(stageSpacing),
                 }
             };
 
             var currentAction = ManiaAction.Key1;
-            for (int i = 0; i < columnCount; i++)
+
+            foreach (var stage in stages)
             {
-                var c = new Column();
-                c.VisibleTimeRange.BindTo(VisibleTimeRange);
+                var drawableStage = new ManiaColumnStage(stage.Columns);
+                drawableStage.VisibleTimeRange.BindTo(VisibleTimeRange);
 
-                c.IsSpecial = isSpecialColumn(i);
-                c.Action = c.IsSpecial ? ManiaAction.Special : currentAction++;
+                listColumnStages.Add(drawableStage);
+                AddNested(drawableStage);
 
-                topLevelContainer.Add(c.TopLevelContainer.CreateProxy());
+                for (int i = 0; i < stage.Columns; i++)
+                {
+                    var c = new Column
+                    {
+                        //c.Action = c.IsSpecial ? ManiaAction.Special : currentAction++;

```

Comment:
This isn't going to work - special columns are broken right now. Mania's Key1 starts at value = 10 to allow for, e.g. `ManiaAction.Special + 1` for the second special column.

Reply:
or just cancel ManiaAction.Special?
Because in co-op mode need two Special in ManiaAction,
and user need to reset the setting when SpecialColumnPosition is left or right.

Reply:
Not sure what you mean by cancel? Each stage should get a special column by doing something like `ManiaAction.Special + stageIndex`